File: audio_metrics.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <sndfile.h>
#include <complex>
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace std;

// Calculate Mean Squared Error
double calculateMSE(const vector<double>& original, const vector<double>& compressed) {
    double sum = 0.0;
    for (size_t i = 0; i < original.size(); ++i) {
        double diff = original[i] - compressed[i];
        sum += diff * diff;
    }
    return sum / original.size();
}

// Calculate PSNR
double calculatePSNR(double mse) {
    if (mse == 0) return INFINITY;
    return 10.0 * log10((32767.0 * 32767.0) / mse);
}

// Calculate SNR
// SNR = 10 * log10(sum(x^2) / sum((x-x_hat)^2))
double calculateSNR(const std::vector<double>& original, const std::vector<double>& compressed) {
    double signalEnergy = 0.0;
    double noiseEnergy = 0.0;
    for (size_t i = 0; i < original.size(); ++i) {
        signalEnergy += original[i] * original[i];
        double noise = original[i] - compressed[i];
        noiseEnergy += noise * noise;
    }
    if (noiseEnergy == 0) return INFINITY;
    return 10.0 * log10(signalEnergy / noiseEnergy);
}

// Calculate compression ratios
struct CompressionRatios {
    double wavRatio;      // ratio between input.wav and output.wav
    double actualRatio;   // ratio between input.wav and compressed.bin
};

CompressionRatios calculateCompressionRatios(const string& inputFile, const string& outputFile) {
    ifstream input(inputFile, ios::binary | ios::ate);
    ifstream output(outputFile, ios::binary | ios::ate);
    ifstream compressed("compressed.bin", ios::binary | ios::ate);
    
    CompressionRatios ratios = {0.0, 0.0};
    
    if (!input) {
        cout << "Warning: Cannot open input file for size calculation" << endl;
        return ratios;
    }
    
    double inputSize = input.tellg();
    
    if (output) {
        double outputSize = output.tellg();
        ratios.wavRatio = inputSize / outputSize;
    }
    
    if (compressed) {
        double compressedSize = compressed.tellg();
        ratios.actualRatio = inputSize / compressedSize;
    } else {
        cout << "Warning: compressed.bin not found" << endl;
    }
    
    return ratios;
}

// Normalize audio to [-1, 1]
vector<double> normalizeAudio(const vector<int16_t>& audio) {
    vector<double> normalized(audio.size());
    for (size_t i = 0; i < audio.size(); ++i) {
        normalized[i] = audio[i] / 32768.0;
    }
    return normalized;
}

// FFT implementation for real signals
void fft(const std::vector<double>& in, std::vector<std::complex<double>>& out) {
    size_t N = in.size();
    out.resize(N);
    if (N == 0) return;
    if ((N & (N - 1)) != 0) throw std::runtime_error("FFT input size must be a power of 2");
    // Bit reversal
    size_t j = 0;
    for (size_t i = 0; i < N; ++i) {
        if (i < j) std::swap(out[i], out[j]);
        out[i] = in[i];
        size_t m = N >> 1;
        while (m >= 1 && j >= m) {
            j -= m;
            m >>= 1;
        }
        j += m;
    }
    // Danielson-Lanczos
    for (size_t s = 1; s <= static_cast<size_t>(log2(N)); ++s) {
        size_t m = 1 << s;
        std::complex<double> wm = std::exp(std::complex<double>(0, -2.0 * M_PI / m));
        for (size_t k = 0; k < N; k += m) {
            std::complex<double> w = 1;
            for (size_t j = 0; j < m / 2; ++j) {
                std::complex<double> t = w * out[k + j + m / 2];
                std::complex<double> u = out[k + j];
                out[k + j] = u + t;
                out[k + j + m / 2] = u - t;
                w *= wm;
            }
        }
    }
}

// Calculate THD (proper version)
double calculateTHD(const std::vector<double>& signal, double sampleRate) {
    // Zero-pad to next power of 2
    size_t N = 1;
    while (N < signal.size()) N <<= 1;
    std::vector<double> padded = signal;
    padded.resize(N, 0.0);
    std::vector<std::complex<double>> spectrum(N);
    fft(padded, spectrum);
    // Compute magnitude spectrum
    std::vector<double> mag(N / 2);
    for (size_t i = 0; i < N / 2; ++i) {
        mag[i] = std::abs(spectrum[i]) / (N / 2);
    }
    // Find fundamental frequency bin (ignore DC)
    size_t fundamentalBin = 1;
    double maxMag = mag[1];
    for (size_t i = 2; i < N / 2; ++i) {
        if (mag[i] > maxMag) {
            maxMag = mag[i];
            fundamentalBin = i;
        }
    }
    // RMS of fundamental
    double V1 = mag[fundamentalBin] / std::sqrt(2.0);
    // RMS of harmonics (2nd to 5th)
    double sumHarmonics2 = 0.0;
    int maxHarmonic = 5;
    for (int h = 2; h <= maxHarmonic; ++h) {
        size_t bin = fundamentalBin * h;
        if (bin < N / 2) {
            double Vh = mag[bin] / std::sqrt(2.0);
            sumHarmonics2 += Vh * Vh;
        }
    }
    double thd = (V1 > 0) ? std::sqrt(sumHarmonics2) / V1 : 0.0;
    return thd;
}

void analyzeAudio(const string& inputFile, const string& outputFile) {
    cout << "Opening files..." << endl;
    
    // Open input file
    SF_INFO inputInfo = {};
    SNDFILE* inFile = sf_open(inputFile.c_str(), SFM_READ, &inputInfo);
    if (!inFile) {
        throw runtime_error("Cannot open input file: " + string(sf_strerror(nullptr)));
    }
    
    // Open output file
    SF_INFO outputInfo = {};
    SNDFILE* outFile = sf_open(outputFile.c_str(), SFM_READ, &outputInfo);
    if (!outFile) {
        sf_close(inFile);
        throw runtime_error("Cannot open output file: " + string(sf_strerror(nullptr)));
    }
    
    cout << "Reading audio data..." << endl;
    
    // Read audio data
    vector<int16_t> inputAudio(inputInfo.frames * inputInfo.channels);
    vector<int16_t> outputAudio(outputInfo.frames * outputInfo.channels);
    
    sf_read_short(inFile, inputAudio.data(), inputAudio.size());
    sf_read_short(outFile, outputAudio.data(), outputAudio.size());
    
    sf_close(inFile);
    sf_close(outFile);
    
    cout << "Processing..." << endl;
    
    // Normalize audio
    vector<double> normalizedInput = normalizeAudio(inputAudio);
    vector<double> normalizedOutput = normalizeAudio(outputAudio);
    
    // Calculate metrics
    double mse = calculateMSE(normalizedInput, normalizedOutput);
    double psnr = calculatePSNR(mse);
    double snr = calculateSNR(normalizedInput, normalizedOutput);
    CompressionRatios ratios = calculateCompressionRatios(inputFile, outputFile);
    double thd = calculateTHD(normalizedOutput, outputInfo.samplerate);
    
    // Print results
    cout << "\n=== Audio Quality Analysis Results ===\n";
    cout << "Sample Rate: " << inputInfo.samplerate << " Hz\n";
    cout << "Channels: " << inputInfo.channels << "\n";
    cout << "Duration: " << static_cast<double>(inputInfo.frames) / inputInfo.samplerate << " seconds\n";
    cout << "\nQuality Metrics:\n";
    cout << "MSE: " << mse << "\n";
    cout << "PSNR: " << psnr << " dB\n";
    cout << "SNR: " << snr << " dB\n";
    cout << "WAV Compression Ratio (input.wav:output.wav): " << ratios.wavRatio << ":1\n";
    cout << "Actual Compression Ratio (input.wav:compressed.bin): " << ratios.actualRatio << ":1\n";
    cout << "THD: " << (thd * 100) << "%\n";
}

int main(int argc, char* argv[]) {
    try {
        if (argc != 3) {
            cout << "Usage: " << argv[0] << " <input.wav> <output.wav>\n";
            return 1;
        }
        
        cout << "Starting audio analysis...\n";
        cout << "Input file: " << argv[1] << "\n";
        cout << "Output file: " << argv[2] << "\n\n";
        
        analyzeAudio(argv[1], argv[2]);
        return 0;
    }
    catch (const exception& e) {
        cerr << "Error: " << e.what() << endl;
        return 1;
    }
} 

File: audio_metrics.py
import numpy as np
import soundfile as sf
import os
from typing import Tuple, List
import math
from scipy.fft import fft

def calculate_mse(original: np.ndarray, compressed: np.ndarray) -> float:
    """Calculate Mean Squared Error between original and compressed signals."""
    return np.mean((original - compressed) ** 2)

def calculate_psnr(mse: float) -> float:
    """Calculate Peak Signal-to-Noise Ratio."""
    if mse == 0:
        return float('inf')
    # MAX^2 for normalized audio in [-1, 1] is 1.0 * 1.0 = 1.0
    return 10.0 * math.log10(1.0 / mse)

def calculate_snr(original: np.ndarray, compressed: np.ndarray) -> float:
    """Calculate Signal-to-Noise Ratio."""
    signal_energy = np.sum(original ** 2)
    noise_energy = np.sum((original - compressed) ** 2)
    if noise_energy == 0:
        return float('inf')
    return 10.0 * math.log10(signal_energy / noise_energy)

def calculate_compression_ratios(input_file: str, output_file: str) -> Tuple[float, float]:
    """Calculate compression ratios between input and output files."""
    input_size = os.path.getsize(input_file)
    output_size = os.path.getsize(output_file)
    compressed_size = os.path.getsize('compressed.bin') if os.path.exists('compressed.bin') else 0
    
    wav_ratio = input_size / output_size if output_size > 0 else 0
    actual_ratio = input_size / compressed_size if compressed_size > 0 else 0
    
    return wav_ratio, actual_ratio

def calculate_thd(audio_data, sample_rate):
    """
    Calculate Total Harmonic Distortion (THD) from audio data.
    
    Args:
        audio_data (numpy.ndarray): Audio signal data
        sample_rate (int): Sample rate of the audio
        
    Returns:
        float: THD value as a percentage
    """
    # Convert to float32 for better precision
    audio_data = audio_data.astype(np.float32)
    
    # Normalize the signal
    if np.max(np.abs(audio_data)) > 0:
        audio_data = audio_data / np.max(np.abs(audio_data))
    
    # Perform FFT
    n = len(audio_data)
    if n == 0:
        return 0.0
    fft_result = fft(audio_data)
    fft_magnitude = np.abs(fft_result[:n//2])
    
    # Find the fundamental frequency (highest peak, ignore DC)
    if len(fft_magnitude[1:]) == 0:
        return 0.0
    fundamental_idx = np.argmax(fft_magnitude[1:]) + 1
    
    # Calculate RMS values
    # Fundamental frequency component
    v1 = fft_magnitude[fundamental_idx]
    
    # Harmonic components (2nd to 10th)
    harmonic_sum_sq = 0
    for i in range(2, 11):
        harmonic_idx = fundamental_idx * i
        if harmonic_idx < len(fft_magnitude):
            # Calculate RMS of harmonic and sum squares
            vh = fft_magnitude[harmonic_idx]
            harmonic_sum_sq += vh ** 2
    
    # Calculate THD
    # The formula in the image uses amplitude for V1 and Vh, so we don't divide by sqrt(2) here
    thd = np.sqrt(harmonic_sum_sq) / v1 if v1 > 0 else 0.0
    thd_percentage = thd * 100
    
    return thd_percentage

def analyze_audio(input_file: str, output_file: str) -> None:
    """Analyze audio files and print quality metrics."""
    print("Opening files...")
    
    # Read input file
    input_data, input_samplerate = sf.read(input_file)
    output_data, output_samplerate = sf.read(output_file)
    
    # Convert to mono if stereo and ensure float type for calculations
    if len(input_data.shape) > 1:
        input_data = np.mean(input_data, axis=1)
    if len(output_data.shape) > 1:
        output_data = np.mean(output_data, axis=1)
    
    # Ensure data is float type for calculations if not already (sf.read normally returns float for WAV)
    if not np.issubdtype(input_data.dtype, np.floating):
         input_data = input_data.astype(np.float32)
         if np.max(np.abs(input_data)) > 0:
             input_data = input_data / np.max(np.abs(input_data))

    if not np.issubdtype(output_data.dtype, np.floating):
         output_data = output_data.astype(np.float32)
         if np.max(np.abs(output_data)) > 0:
             output_data = output_data / np.max(np.abs(output_data))


    print("Processing...")
    
    # Calculate metrics
    # Ensure inputs for MSE and SNR are the same length, truncate to shorter if necessary
    min_len = min(len(input_data), len(output_data))
    input_data_trunc = input_data[:min_len]
    output_data_trunc = output_data[:min_len]

    mse = calculate_mse(input_data_trunc, output_data_trunc)
    psnr = calculate_psnr(mse)
    snr = calculate_snr(input_data_trunc, output_data_trunc)
    compression_ratios = calculate_compression_ratios(input_file, output_file)
    # THD calculated using the new formula
    thd_input = calculate_thd(input_data, input_samplerate)
    thd_output = calculate_thd(output_data, output_samplerate)
    
    # Print results
    print("\n=== Audio Quality Analysis Results ===")
    print(f"Sample Rate: {input_samplerate} Hz")
    print(f"Input Channels: {1 if len(input_data.shape) == 1 else input_data.shape[1]}")
    print(f"Output Channels: {1 if len(output_data.shape) == 1 else output_data.shape[1]}")
    print(f"Input Duration: {len(input_data) / input_samplerate:.2f} seconds")
    print(f"Output Duration: {len(output_data) / output_samplerate:.2f} seconds")
    print("\nQuality Metrics:")
    print(f"MSE: {mse:.6f}")
    print(f"PSNR: {psnr:.2f} dB")
    print(f"SNR: {snr:.2f} dB")
    print(f"WAV Compression Ratio (input.wav:output.wav): {compression_ratios[0]:.2f}:1")
    print(f"Actual Compression Ratio (input.wav:compressed.bin): {compression_ratios[1]:.2f}:1")
    print(f"THD (Input): {thd_input:.2f}%")
    print(f"THD (Output): {thd_output:.2f}%")

def main():
    import sys
    
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <input.wav> <output.wav>")
        sys.exit(1)
    
    print("Starting audio analysis...")
    print(f"Input file: {sys.argv[1]}")
    print(f"Output file: {sys.argv[2]}\n")
    
    try:
        analyze_audio(sys.argv[1], sys.argv[2])
    except Exception as e:
        print(f"Error: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()
